plugins/cryptolib/cryptolib.cpp

```diff
@@ -0,0 +1,667 @@
+        throw MakeStringException(-1, "Unsupported hash algorithm '%s' specified", algorithm);
+    }
+}
+
+//-------------------------------------------------------------------------------------------------------------------------------------------
+// PUBLIC KEY SUPPORT
+//-------------------------------------------------------------------------------------------------------------------------------------------
+
+CRYPTOLIB_API void CRYPTOLIB_CALL clInstalledPublicKeyAlgorithms(bool & __isAllResult, size32_t & __lenResult, void * & __result)
+{
+    __isAllResult = false;
+    StringArray algorithms;
+#if OPENSSL_VERSION_NUMBER < 0x10100000L
+    loadDigests();
+#endif
+    OBJ_NAME_do_all_sorted(OBJ_NAME_TYPE_PKEY_METH, my_clAlgorithmsCallback, (void*)&algorithms);
+    stringArrayToECLSetOfString(algorithms, &__result, &__lenResult);
+}
+
+CRYPTOLIB_API void CRYPTOLIB_CALL clSupportedPublicKeyAlgorithms(bool & __isAllResult, size32_t & __lenResult, void * & __result)
+{
+    __isAllResult = false;
+    StringArray algorithms;
+    algorithms.appendList("SHA256", ",");
+    stringArrayToECLSetOfString(algorithms, &__result, &__lenResult);
+}
+
+//signature helper function
+void doPKISign(size32_t & __lenResult, void * & __result,
+        IDigitalSignatureManager * pDSM,
+        size32_t lenInputdata, const void * inputdata)
+{
+    StringBuffer sbSig;
+    if (pDSM->digiSign(sbSig, lenInputdata, inputdata))
+    {
+        __result = CTXMALLOC(parentCtx, sbSig.length());
+        __lenResult = sbSig.length();
+        memcpy(__result, sbSig.str(), __lenResult);
+    }
+    else
+    {
+        throw MakeStringException(-1, "Unable to create Digital Signature");
+    }
+}
+
+void verifyPKIAlgorithm(const char * pkAlgorithm)
+{
+    if (0 != stricmp(pkAlgorithm, "SHA256"))
+        throw MakeStringException(-1, "Unsupported PKI algorithm (%s) specified", pkAlgorithm);
+}
+
+
+//-----------------------------------------------------------------
+//  Simple cache for instances of Digital Signature Managers
+//-----------------------------------------------------------------
+class CDSMCache
+{
+private:
+    CriticalSection signingDSMCacheCS;//guards modifications to the cache map
+    CriticalSection verifyingDSMCacheCS;//guards modifications to the cache map
+    typedef std::map<string, IDigitalSignatureManager *> DSMCache;//DSM instances are never deleted, since we are not notified if a module is destroyed
+    DSMCache signingDSMCache;
+    DSMCache verifyingDSMCache;
+
+public:
+
+    IDigitalSignatureManager * querySigningInstance(const char * algo, const char * privKeyFS, const char * privKeyBuff, const char * passphrase)
+    {
+        VStringBuffer searchKey("%s_%s_%s_%s", algo, privKeyFS, privKeyBuff, passphrase);
+        CriticalBlock block(signingDSMCacheCS);
+        DSMCache::iterator it = signingDSMCache.find(searchKey.str());
+        if (it != signingDSMCache.end())//exists in cache
+        {
+            return (*it).second;
```

Comment:
related to previous comment, could this be a getXX method and return a linked object?, and then they will be destroyed correctly.

Reply:
@jakesmith I am referring to the ECL Module , which is SymmetricEncryption in this case. Gavin told me that when it (an ECL Module) goes out of scope the plugin is not notified, so I  have to keep everything around until termination time

Reply:
Sure, but don't think that means you have to leak them when the c++ module/plugin is unloaded.
Worth checking with @ghalliday, but as per above, I think you should link/own them, and therefore ensure they are release destroyed when the plugin is unloaded